### PR TITLE
[AC-1971]: Use git submodule for uikit-react to reduce the bundle size

### PR DESCRIPTION
**Pull request by AhyoungRyu**
### Key Changes:

- Removed the `@sendbird/uikit-react` dependency from `package.json`.
- Linked the necessary code from [sendbird/uikit-react](https://github.com/sendbird/sendbird-uikit-react) directly into the `packages/uikit/` dir through a Git submodule. <- See `.gitmodules` file
- Updated our build process(Github workflow / Circleci config) to initialize and update the Git submodule, ensuring that the latest version of the code is always used.

So how much of the bundle size would be reduced? 
**AS-IS**
Screenshot 2024-04-29 at 5 55 49 PM

**TO-BE**
Screenshot 2024-04-29 at 5 50 40 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,6 +15,7 @@
     "url": "https://github.com/sendbird/chat-ai-widget.git"
   },
   "scripts": {
+    "install:deps": "npm install --force && git submodule update --init",
```
Comment (AhyoungRyu) [inline package.json]: Why we're using `--force` param here? There's a peer dep setting in `@mojojoejo/vite-plugin-purgecss` like below, so we can't use the latest vite version. So I put `--force` option to ignore the error.
Screenshot 2024-04-29 at 5 56 51 PM

File: src/components/BotProfileImage.tsx [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import styled from 'styled-components';
 
 import { getColorBasedOnSaturation } from '../colors';
-import { ReactComponent as Icon } from '../icons/bot-profile-image-small.svg';
```
Comment (AhyoungRyu) [inline src/components/BotProfileImage.tsx:4]: Why this format have to be changed? 
- Verbose to use
- It causes a conflict with UIKit svg import type and throwing an error 🥲

File: package.json (hunk 1 above)
Comment (bang9) [reply package.json]: Could you test if it also works using the Rollup plugin instead of the Vite plugin?
- https://v1.purgecss.com/with-rollup

Comment (AhyoungRyu) [reply package.json]: @bang9  Hmm it seems it also conflicts with other packages and doesn't give a happy result. Let me split this purgecss task and do it separately :) 

I'll put the purgecss related stuff.

Comment (bang9) [reply package.json]: 😭

Comment (bang9) [reply package.json]: If the reduction in bundle size for each Vite version is not significantly noticeable, it might be better to maintaining version compatibility

Comment (AhyoungRyu) [reply package.json]: Yeah it seems there's no diff between 5 and 6 in terms of the bundle size. So I just used the latest version of Vite which is 5.2.10 and updated other plugins' version accordingly. https://github.com/sendbird/chat-ai-widget/pull/180/commits/124f51ec746820db2747e3f4bb9e7858bd1e7dbd

File: .eslintrc.cjs [hunk 3]
```diff
@@ -2,38 +2,40 @@ module.exports = {
   root: true, // This fixes issue when running lint fix command: https://stackoverflow.com/questions/55060228/eslint-couldnt-find-the-plugin-eslint-plugin-typescript-eslint
   env: { browser: true, es2020: true },
   parser: '@typescript-eslint/parser',
-  parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
-  plugins: ['import', 'styled-components-a11y'],
+  parserOptions: {
+    ecmaVersion: 'latest',
+    sourceType: 'module',
+    ecmaFeatures: { jsx: true },
```
Comment (chohongm) [inline .eslintrc.cjs:8]: jsx: ECMScript 규격의 [JSX](https://facebook.github.io/jsx/) 사용 여부
^ is this relevant to reducing bundle size?

File: .eslintrc.cjs [hunk 4]
```diff
@@ -2,38 +2,40 @@ module.exports = {
...
+  },
+  plugins: ['import', '@typescript-eslint', 'styled-components-a11y'],
```
Comment (chohongm) [inline .eslintrc.cjs]: why we need @typescript-eslint?

File: src/components/MessageDataContent.tsx [hunk 5]
```diff
@@ -1,9 +1,9 @@
 import styled from 'styled-components';
 
 import { useConstantState } from '../context/ConstantContext';
-import { ReactComponent as ChevronRightIcon } from '../icons/icon-chevron-right.svg';
-import { ReactComponent as EllipsisIcon } from '../icons/icon-ellipsis.svg';
-import { ReactComponent as MessageBubbleIcon } from '../icons/icon-message-bubble.svg';
```
Comment (chohongm) [inline src/components/MessageDataContent.tsx:6]: im not sure why we had to use
{ ReactComponent as xxx } format in the first place

File: tsconfig.json [hunk 6]
```diff
@@ -20,7 +20,10 @@
     "noUnusedLocals": true,
     "noUnusedParameters": true,
     "noFallthroughCasesInSwitch": true,
-    "allowSyntheticDefaultImports": true
+    "allowSyntheticDefaultImports": true,
+    "paths": {
+      "@uikit/*": ["./packages/uikit/src/*"],
+    }
```
Comment (chohongm) [inline tsconfig.json:26]: what is path for?

File: vite.config.ts [hunk 7]
```diff
@@ -12,12 +18,29 @@ export default defineConfig({
...
+      },
+    }),
     dts(),
+    visualizer({
+      filename: './dist/report.html',
+      brotliSize: true,
+    }),
+    // pluginPurgeCss({
+    //   variables: true,
+    // }),
```
Comment (chohongm) [inline vite.config.ts]: lets remove unused

File: vite.config.ts [hunk 8]
```diff
@@ -12,12 +18,29 @@ export default defineConfig({
...
   ],
-  define: {
-    APP_VERSION: JSON.stringify(process.env.npm_package_version),
+  resolve: {
+    alias: [
+      {
+        find: '@uikit',
+        replacement: path.resolve(__dirname, 'packages/uikit/src'),
+      },
+    ],
```
Comment (chohongm) [inline vite.config.ts:39]: what is resolve for?

File: vite.config.ts [hunk 9]
```diff
@@ -26,19 +49,17 @@ export default defineConfig({
...
       formats: ['es', 'umd'],
-      fileName: (format) => `index.${format}.js`
+      fileName: (format) => `index.${format}.js`,
     },
     rollupOptions: {
-      external: [
-        'react',
-        'react-dom',
-      ],
+      plugins: [terser()],
```
Comment (chohongm) [inline vite.config.ts:51]: great idea! for curiousity, how much in % is size reduced by terser?

File: .gitmodules [hunk 10]
```diff
@@ -0,0 +1,4 @@
+[submodule "packages/uikit"]
+	path = packages/uikit
+	url = git@github.com:sendbird/sendbird-uikit-react.git
+	branch = ai-widget/experimental
```
Comment (chohongm) [inline .gitmodules:4]: what is this file for?

**Review by chohongm — COMMENTED** [review]
Is terser the only change that reduced bundle size?

File: tsconfig.json (hunk 6 above)
Comment (AhyoungRyu) [reply tsconfig.json:26]: Related to the thing I mentioned in the key change. 

>what is path for?
 
To enable this `@uikit/***`.

File: vite.config.ts (hunk 8 above)
Comment (AhyoungRyu) [reply vite.config.ts:39]: Also related to https://github.com/sendbird/chat-ai-widget/pull/180#discussion_r1584204401

File: .gitmodules (hunk 10 above)
Comment (AhyoungRyu) [reply .gitmodules:4]: Related to the things I mentioned in the PR description. 
Screenshot 2024-04-30 at 3 39 04 PM

Let me know if you need more detailed explanation :)

File: tsconfig.json (hunk 6 above)
Comment (chohongm) [reply tsconfig.json:26]: so this is necessary when using submodule?

File: .eslintrc.cjs (hunk 3 above)
Comment (AhyoungRyu) [reply .eslintrc.cjs:8]: Size reduce related? Not really but added in terms of better practice? \cc @bang9

File: vite.config.ts (hunk 9 above)
Comment (AhyoungRyu) [reply vite.config.ts:51]: - With
Screenshot 2024-04-30 at 3 44 31 PM

- Without
Screenshot 2024-04-30 at 3 44 05 PM

**Comment by AhyoungRyu** [comment]
>Is terser the only change that reduced bundle size?

Yes :)

File: vite.config.ts (hunk 7 above)
Comment (AhyoungRyu) [reply vite.config.ts]: I commented this out since we're talking about the purgecss in here https://sendbird.slack.com/archives/C05QD5M7XDH/p1714457280742949?thread_ts=1714383967.760279&cid=C05QD5M7XDH. TBD yet 🤔

File: .eslintrc.cjs (hunk 3 above)
Comment (chohongm) [reply .eslintrc.cjs:8]: before adding this was there an issue?

File: vite.config.ts (hunk 9 above)
Comment (chohongm) [reply vite.config.ts:51]: oh it is less than what I expected (>= 20%)

**Review by chohongm — APPROVED** [review]
LGTM! Great job~

File: .eslintrc.cjs (hunk 4 above)
Comment (AhyoungRyu) [reply .eslintrc.cjs]: It seems we don't need to specify this in both `plugins` and `extends`. So I removed it from plugins https://github.com/sendbird/chat-ai-widget/pull/180/commits/dccade2231ce2c82258e93fbf8fbdaae43e576ae

File: .eslintrc.cjs (hunk 3 above)
Comment (bang9) [reply .eslintrc.cjs:8]: I directly linked UIKit and set up path aliases, ensuring that files are processed correctly in ESLint as well.
But I can't remember why `ecmaFeatures: { jsx: true }` was enabled. 🙄

File: package.json [hunk 11]
```diff
@@ -29,37 +31,45 @@
...
     "eslint-plugin-react-hooks": "^4.6.0",
     "eslint-plugin-styled-components-a11y": "^2.1.32",
     "prettier": "2.8.8",
+    "prettier-eslint": "^15.0.1",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "rollup-plugin-terser": "^7.0.2",
+    "rollup-plugin-visualizer": "^5.12.0",
     "tsc-silent": "^1.2.2",
+    "tslint-config-prettier": "^1.18.0",
```
Comment (bang9) [inline package.json]: We are using eslint-typescript instead of the deprecated tslint. It seems like we can remove it.

**Review by bang9 — APPROVED** [review]

File: package.json (hunk 11 above)
Comment (AhyoungRyu) [reply package.json]: Yeah removed it too https://github.com/sendbird/chat-ai-widget/pull/180/commits/2d02b80d533f7e74659d7736f05f9321f7ae5877

File: vite.config.ts (hunk 7 above)
Comment (AhyoungRyu) [reply vite.config.ts]: Removed it https://github.com/sendbird/chat-ai-widget/pull/180/commits/26d1e7283ae06399b12a079cd4ce2e8d2bb02f13 :)
